lowcar/devices/KoalaBear/KoalaBear.cpp

```diff
@@ -5,358 +5,346 @@
 
-	//this->pid = new PID(0.0, 1.0, 0.0, 0.0, (float) millis());
-	this->pid_enabled = (uint8_t *) calloc(2, sizeof(uint8_t));
-
-	this->prev_led_time = millis();
-	this->curr_led_mtr = MTRA;
-
-	this->led = new LEDKoala();
-	// TODO: Initialize encoder and PID
+// TODO: write a descriptive comment here once we know this works
+static void handle_enc_a_tick() {
+    if (digitalRead(AENC2)) {
+        enc_a++;
+    } else {
+        enc_a--;
+    }
 }
 
-size_t KoalaBear::device_read (uint8_t param, uint8_t *data_buf) {
-	float *float_buf = (float *) data_buf;
-	bool *bool_buf = (bool *) data_buf;
-
-	switch (param) {
-
-		case DUTY_CYCLE_A:
-			float_buf[0] = this->desired_speeds[MTRA];
-			return sizeof(float);
-
-		case DEADBAND_A:
-			float_buf[0] = this->deadbands[MTRA];
-			return sizeof(float);
-
-		case CURRENT_A:
-			// TODO
-			return sizeof(float);
-
-		case PID_ENABLED_A:
-			float_buf[0] = this->pid_enabled[MTRA];
-			return sizeof(uint8_t);
-
-		//TODO: ask PID controller for motor A for KP
-		case PID_KP_A:
-			return sizeof(float);
-
-		//TODO: ask PID controller for motor A for KI
-		case PID_KI_A:
-			return sizeof(float);
-
-		//TODO: ask PID controller for motor A for KD
-		case PID_KD_A:
-			return sizeof(float);
-
-		case ENC_A:
-			//float_buf[0] = this->pid->readPos();
-			return sizeof(float);
-
-		// Params for Motor B
-
-		case DUTY_CYCLE_B:
-			float_buf[0] = this->desired_speeds[MTRB];
-			return sizeof(float);
-
-		case DEADBAND_B:
-			float_buf[0] = this->deadbands[MTRB];
-			return sizeof(float);
-
-		case CURRENT_B:
-			// TODO
-			return sizeof(float);
-
-		case PID_ENABLED_B:
-			data_buf[0] = this->pid_enabled[MTRB];
-			return sizeof(uint8_t);
-
-		//TODO: ask PID controller for motor B for KP
-		case PID_KP_B:
-			return sizeof(float);
+// TODO: write a descriptive comment here once we know this works
+static void handle_enc_b_tick() {
+    if (digitalRead(BENC2)) {
+        enc_b++;
+    } else {
+        enc_b--;
+    }
+}
 
-		//TODO: ask PID controller for motor B for KI
-		case PID_KI_B:
-			return sizeof(float);
+//*********************************** MAIN KOALABEAR CODE ***********************************//
 
-		//TODO: ask PID controller for motor B for KD
-		case PID_KD_B:
-			return sizeof(float);
+KoalaBear::KoalaBear() : Device(DeviceType::KOALA_BEAR, 13) {
+    // initialize motor values    
+    this->target_speed_a = this->target_speed_b = 0.0;
+    this->deadband_a = this->deadband_b = 0.05;
+    
+    // initialize encoders
+    enc_a = enc_b = 0.0;
+    attachInterrupt(digitalPinToInterrupt(AENC1), handle_enc_a_tick, RISING); // set interrupt service routines for encoder pins
+    attachInterrupt(digitalPinToInterrupt(BENC1), handle_enc_b_tick, RISING);
+    
+    // initialize PID controllers
+    this->pid_a = new PID();
+    this->pid_b = new PID();
+    this->pid_enabled_a = this->pid_enabled_b = FALSE; // TODO: change to true when PID written
```

Comment:
why is FALSE not `false`? Also, they should default to disabled and the students should turn them on when they wish.

Reply:
To answer the first question, it's because `pid_enabled_a` is a boolean, but in C++, true/false is not necessarily 1 and 0. So if you use the C++ true/false values, when you send them over serial to `dev_handler`, the true value ends up being a random positive number between 1 and 255 since C doesn't have a builtin boolean type. The solution is to define two constants, TRUE as 1 and FALSE as 0, so that when we send them to `dev_handler` they can be treated as true/false values in C.

To answer the second point, I talked a bit about this with Jesse and Jonathan. They both believed that our students likely won't be advanced enough to be able to tune the PID controllers themselves. So, by default these will be enabled and the KP, KI, and KD values will have default values that we find for them. They can change the tuning constants and/or disable the PID control if they wish, but the PID control is mostly for us to try and create a better experience for the students (robots actually drive straight in autonomous, etc.)

Reply:
even if that were the case, you probably wouldn't want the pid to run during teleop. also right now theres no way for them to  give a setpoint to the pid controller in the device params, which is pretty important.